src/tools/singlejar/port.cc

```diff
@@ -0,0 +1,36 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef _WIN32
+
+#include "src/tools/singlejar/port.h"
+
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+#include <windows.h>
+
+ptrdiff_t pread(int fd, void *buf, size_t count, ptrdiff_t offset) {
+  DWORD ret = -1;
+  HANDLE hFile = reinterpret_cast<HANDLE>(_get_osfhandle(fd));
+  if (hFile) {
+    OVERLAPPED overlap;
+    memset(&overlap, 0, sizeof(OVERLAPPED));
+    overlap.Offset = offset;
+    ::ReadFile(hFile, buf, count, &ret, &overlap);
```

Comment:
`ReadFile` will only use `lpOverlapped` if the file was opened with `FILE_FLAG_OVERLAPPED`. Is that that case here?

Reply:
Internal reviewer comment:

What happens if ReadFile fails?

Reply:
From [MSDN](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-readfile), `&ret` is not guaranteed to be set correctly on error. `ReadFile` will return `FALSE` on error, so we check that and just return `-1` according to [POSIX specification](https://www.gnu.org/software/libc/manual/html_node/I_002fO-Primitives.html).